[RDPF - 64] 회원탈퇴 useCase 작성

kma952727 opened the pull request:
## What is this PR? 👀

delete "/v1/user" API 작성, 사용자는 api호출 시 회원탈퇴(soft delete)

탈퇴 성공 시 http status 202
탈퇴 실패 시 http status 401을 반환

- Issue ticket number: RDPF - 64

## Changes ✏️

1. perfume-api/src/main/java/io/perfume/api/auth/application/service/AuthenticationTokenService.java에 간단한 개선사항 TODO 작성
2. User Model에 softDelete() 메서드 작성

## Test checklist 🧪

<!-- 기능 추가 시 작성한 테스트 코드 목록을 작성해주세요. -->

- [ ]

kma952727 committed:
[RDPF - 64] 회원탈퇴 useCase 작성

1. delete "/v1/users" API 작성, 사용자는 api호출 시 회원탈퇴(soft delete)

kma952727 committed:
[RDPF-64] UsersSupportController의 path prefix "/v1/users"에서 "/v1/user" 로 수정

1. 해당 컨트롤러의 기능들이 모두 자신의 자원에 접근하는 기능만 구현, users가 path에 포함되면은 안된다.

kma952727 committed:
[RDPF-64]  "/email/reset-passowrd", "/find-username" api에서 "/user" prefix 제거

1. "/email/reset-passowrd", "/find-username" api가 자신의 자원에 접근하는 개념이 아님, "/user" prefix 제거

kma952727 committed:
Merge branch 'develop' into RDPF-64

oliviarla commented on perfume-api/src/main/java/io/perfume/api/user/application/exception/FailedLeaveException.java:
```diff
@@ -0,0 +1,12 @@
+import io.perfume.api.base.LogLevel;
+import org.springframework.http.HttpStatus;
+
+public class FailedLeaveException extends CustomHttpException {
```
이름을 FailedToLeaveException 로 변경하는게 좋을 것 같습니다!

kma952727 replied:
수정하겠습니다!

oliviarla commented on perfume-api/src/main/java/io/perfume/api/user/application/service/SupportUserService.java:
```diff
@@ -17,21 +20,22 @@
      * 1. 이메일과 아이디를 알면 누구나 다른 사용자의 암호를 수정할 수 있음
      * 2. "새로운 패스워드 할당" 이메일 템플릿 정하기
      */
-    @Transactional
```
여기는 왜 삭제되었나요?🧐

kma952727 replied:
정말 꼼꼼하시네요! 수정했습니다!

oliviarla commented on perfume-api/src/main/java/io/perfume/api/user/adapter/in/http/UsersSupportController.java:
```diff
@@ -37,4 +39,9 @@ public ResponseEntity<Object> sendUsernameToUsersEmail(@RequestParam @Email Stri
     }
 
+    @DeleteMapping("/user")
+    public ResponseEntity leaveUser(@RequestHeader(name = "Authorization") String accessToken) {
```
현재 access token이 헤더로 들어오고 있지 않습니다..!
직접 access token을 받는 대신에 @AuthenticationPrincipal로 SecurityContext 내의 UserPrincipal 구현체를 받아오는게 좋을 것 같아요!

oliviarla commented on perfume-api/src/main/java/io/perfume/api/user/application/port/in/LeaveUserUseCase.java:
```diff
@@ -0,0 +1,11 @@
+     *
+     * @param accessToken 삭제를 위한 User 데이터 조회에 필요한 값이 담겨있다.
+     */
+    void leave(String accessToken);
```
accessToken 대신 user의 PK를 인자로 받도록 변경이 필요합니당!

oliviarla commented on perfume-api/src/main/java/io/perfume/api/user/application/service/SupportUserService.java:
```diff
@@ -54,4 +58,33 @@ public String findEncryptedUsername(String email) {
+            // TODO 'userId'같은 상수값이 사용할 때 마다 정의해서 사용중, 상수값을 모아서 관리하고싶음
+            userId = jsonWebTokenGenerator.getClaim(accessToken, "usdId", Long.class);
+
+            User user = userQueryRepository.loadUser(userId).orElseThrow(NotFoundUserException::new);
+
+            LocalDateTime now = LocalDateTime.now();
+
+            user.softDelete(now);
+
+            userRepository.save(user);
```
@dygma0 @Big-Cir97 @kma952727 
코딩 스타일에 관한 질문입니다.
저는 이 부분을 userRepository의 softDelete() 메소드 하나로 묶어서 처리하면 변경 감지로 처리할 수 있겠다 생각이 들었는데, DDD 를 따른다고 생각하면 이 방식도 괜찮아보이네요.

각자의 의견이 궁금합니다 🙌🏻

kma952727 replied:
Adapter에 delete 로직이 들어가면 현재 구현한 Adapter를 걷어냈을 때 softDelete() 구현로직이 사라지지 않는게 아닐까? 생각이 듭니다.
다른 분들 의견도 궁금합니다!

oliviarla reviewed (changes requested):
고생하셨습니다 ! 리뷰 달아보았어요
그리고 회원 탈퇴 용어를 leave vs delete 중에 leave로 선택하신 이유가 따로 있나요?

kma952727 committed:
[RDPF-64] 리뷰 개선안 적용

1. UsersSupportController.leaveUser()에서 삭제 대상 식별자를 받아오는 방법 변경 (Header -> SecurityContextHolder)

2. 예외 class 이름 변경 (FailedLeaveException -> FailedToLeaveException)

3. 1번에서 메서드에서 호출하는 useCase 파라미터 타입 변경 (access token -> userId)

kma952727 committed:
[RDPF-64] repush

1. 사유 : 테스트 실패

kma952727 committed:
[RDPF-64] @AuthenticationPrincipal로 받아오는 인증된 사용자 클래스 변경

1. UserPrincipal -> User

kma952727 committed:
[RDPF-64] @AuthenticationPrincipal 참조 클래스 변경

1.  io.perfume.api.user.domain.User -> org.springframework.security.core.userdetails.User
